File: src/omnia/data/structure/mutable/MaskingSet.java
package omnia.data.structure.mutable;

import static java.util.Objects.requireNonNull;

import java.util.Iterator;
import java.util.stream.Stream;

class MaskingSet<E, J extends java.util.Set<E>> implements MutableSet<E> {

  private final J javaSet;

  MaskingSet(J javaSet) {
    this.javaSet = requireNonNull(javaSet);
  }

  protected final J javaSet() {
    return javaSet;
  }

  @Override
  public void add(E element) {
    javaSet.add(element);
  }

  @Override
  public boolean remove(Object element) {
    return javaSet.remove(element);
  }

  @Override
  public void clear() {
    javaSet.clear();
  }

  @Override
  public Iterator<E> iterator() {
    return javaSet.iterator();
  }

  @Override
  public boolean contains(Object element) {
    return javaSet.contains(element);
  }

  @Override
  public boolean isPopulated() {
    return !javaSet.isEmpty();
  }

  @Override
  public int count() {
    return javaSet.size();
  }

  @Override
  public Stream<E> stream() {
    return javaSet.stream();
  }
}

File: src/omnia/data/structure/mutable/MaskingSet.kt
package omnia.data.structure.mutable

import omnia.data.structure.Collection

open class MaskingSet<E : Any, J : kotlin.collections.MutableSet<E>>(private val kotlinSet: J) :
  MutableSet<E> {

  protected fun javaSet(): J {
    return kotlinSet
  }

  override fun add(item: E) {
    kotlinSet.add(item)
  }

  override fun addAll(items: Collection<out E>) {
    kotlinSet.addAll(items)
  }

  override fun removeUnknownTyped(item: Any?): Boolean {
    return kotlinSet.remove(item)
  }

  override fun clear() {
    kotlinSet.clear()
  }

  override fun iterator(): Iterator<E> {
    return kotlinSet.iterator()
  }

  override fun containsUnknownTyped(item: Any?): Boolean {
    return kotlinSet.contains(item)
  }

  override val isPopulated: Boolean
    get() = kotlinSet.isNotEmpty()

  override fun count(): Int {
    return kotlinSet.size
  }
}